docs/proposals/memory-management.md

```diff
@@ -0,0 +1,135 @@
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/resource-qos.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Memory Management in Kubernetes
+
+**Author**: Vishnu Kannan (@vishh)
+
+**Status**: Draft proposal; 
+
+*This document captures the current state of memory handling in kubernetes nodes and suggests some features for improving node reliability and performance*
+
+####Current Model
+Kubelet (via cAdvisor) exposes the total memory capacity on the node to master. System daemons like docker, kube-proxy, and kubelet are moved to a static cgroup. Most of the other system services, excepting init, are moved to a `/system` cgroup. Either no resource restrictions are in place for these system level cgroups or even if they do, they are set to high values (docker is limited to 70% of the total capacity). Hence it is difficult to predict the amount of memory available for pods on any given node.
+
+Cgroups are used to place memory limits to containers. It is probably reasonable to expect that most users don't set memory limits. Docker is used to track OOM kills in containers. Docker does not cleanup the entire container when an OOM kill is encountered. If the pid `1` in the container gets killed, the container is cleaned up immediately. If not, the container has the opportunity to handle the OOM kill gracefully (e.x. stress container). Kubelet currently surfaces the OOM kill activity to users via Kubernetes APIs. 
+
+Quality of Service was recently introduced. There are three classes of service:
+
+1. Guaranteed - Memory requests are guaranteed
+
+2. Burstable - Memory limits guaranteed up to user request
+
+3. Best Effort - No guaranty - memory availability can change over time.
+
+The scheduler does not take into account the total usage on the node. 
+Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+Kubelet (via cadvisor) can track the memory usage of system daemons and pods. It does not take any action based on usage. 
+
+The kernel OOM killer is expected to be enabled. oom_score_adj is used to protect guaranteed jobs and system daemons. Burstable jobs are given a oom_score_adj that will result in them having a oom_score of `1000` as soon as they hit their original request, making them a likely candidate for being killed by the oom killer. Best effort tasks always have a `oom_score_adj` of `1000` which makes them the most likely candidates to be killed. The kernel usually chooses the highest memory consuming process ([along with a few other policies](https://lwn.net/Articles/317814/))
```

Comment:
give an example of what is a system daemon (you presumably don't mean docker, kubelet, or kube-proxy since you mentioned them earlier)

"The kernel usually chooses the highest memory consuming process" I assume you mean "among the processes with the same oom_score" -- say that